contracts/UniswapWrapper.sol

```diff
@@ -0,0 +1,173 @@
+pragma solidity ^0.5.8;
+
+import "@aragon/court/contracts/lib/os/IsContract.sol";
+import "@aragon/court/contracts/lib/os/ERC20.sol";
+import "@aragon/court/contracts/lib/os/SafeERC20.sol";
+import "@aragon/court/contracts/standards/ERC900.sol";
+import "./lib/uniswap/interfaces/IUniswapExchange.sol";
+import "./lib/uniswap/interfaces/IUniswapFactory.sol";
+
+
+contract UniswapWrapper is IsContract {
+    using SafeERC20 for ERC20;
+
+    string private constant ERROR_TOKEN_NOT_CONTRACT = "UW_TOKEN_NOT_CONTRACT";
+    string private constant ERROR_REGISTRY_NOT_CONTRACT = "UW_REGISTRY_NOT_CONTRACT";
+    string private constant ERROR_UNISWAP_FACTORY_NOT_CONTRACT = "UW_UNISWAP_FACTORY_NOT_CONTRACT";
+    string private constant ERROR_ZERO_AMOUNT = "UW_ZERO_AMOUNT";
+    string private constant ERROR_TOKEN_TRANSFER_FAILED = "UW_TOKEN_TRANSFER_FAILED";
+    string private constant ERROR_TOKEN_APPROVAL_FAILED = "UW_TOKEN_APPROVAL_FAILED";
+    string private constant ERROR_ETH_REFUND = "UW_ETH_REFUND";
+    string private constant ERROR_TOKEN_REFUND = "UW_TOKEN_REFUND";
+    string private constant ERROR_UNISWAP_UNAVAILABLE = "UW_UNISWAP_UNAVAILABLE";
+
+    bytes32 internal constant ACTIVATE_DATA = keccak256("activate(uint256)");
+
+    ERC20 public bondedToken;
+    ERC900 public registry;
+    IUniswapFactory public uniswapFactory;
+
+    constructor(ERC20 _bondedToken, ERC900 _registry, IUniswapFactory _uniswapFactory) public {
+        require(isContract(address(_bondedToken)), ERROR_TOKEN_NOT_CONTRACT);
+        require(isContract(address(_registry)), ERROR_REGISTRY_NOT_CONTRACT);
+        require(isContract(address(_uniswapFactory)), ERROR_UNISWAP_FACTORY_NOT_CONTRACT);
+
+        bondedToken = _bondedToken;
+        registry = _registry;
+        uniswapFactory = _uniswapFactory;
+    }
+
+    /**
+    * @dev This function must be triggered by the external token approve-and-call fallback.
+    *      It will pull the approved tokens and convert them in Uniswap, and activate the converted tokens into a
+    *      jurors registry instance of an Aragon Court.
+    * @param _from Address of the original caller (juror) converting and activating the tokens
+    * @param _amount Amount of external tokens to be converted and activated
+    * @param _token Address of the external token triggering the approve-and-call fallback
+    * @param _data If non-empty it will signal token activation in the registry
+    */
+    function receiveApproval(address _from, uint256 _amount, address _token, bytes calldata _data) external {
+        bool activate;
+        uint256 minTokens;
+        uint256 minEth;
+        uint256 deadline;
+        bytes memory data = _data;
+        assembly {
+            activate := mload(add(data, 0x20))
```

Comment:
I think we may want to double check the length of `_data` first before we assume that it's all available; otherwise we may end up with garbage.

Reply:
On this, if for example deadline is 0 because there's no data, would it work or would it always revert because time is greater than deadline?

Reply:
Yes, Uniswap parameters are passed to it as they are. So same for `minEth` and `minTokens` that need to be `> 0` too: https://github.com/Uniswap/contracts-vyper/blob/master/contracts/uniswap_exchange.vy#L272